File: LeetCode/36ValidSudoku/ValidSudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        // Arrays -> Time:O(n^2), Space:O(n^2) n=SideOfBoard
        vector<vector<int> > rows(9, vector<int>(9, 0));
        vector<vector<int> > cols(9, vector<int>(9, 0));
        vector<vector<int> > boxes(9, vector<int>(9, 0));
        int box_index, num;
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] != '.') {
                    box_index = (i/3 + j/3) + (i/3 * 2);
                    num = board[i][j] - '0' - 1;
                    if (rows[i][num] || cols[num][j] || boxes[box_index][num]) {
                        return false;
                    }
                    rows[i][num] = 1;
                    cols[num][j] = 1;
                    boxes[box_index][num] = 1;                    
                }
            }
        }
        return true;

        // // Single pass -> Time:O(n^2), Space:O(n^2) n=SideOfBoard
        // map<string, set<char> > m;
        // int box_index;
        // bool num_in_row, num_in_col, num_in_box;
        // for (int i = 0; i < 9; i++) {
        //     set<char> a;
        //     m["R"+to_string(i)] = a;
        //     set<char> b;
        //     m["C"+to_string(i)] = b;
        //     set<char> c;
        //     m["B"+to_string(i)] = c;
        // }
        // for (int i = 0; i < 9; i++) {
        //     for (int j = 0; j < 9; j++) {
        //         if (board[i][j] != '.') {
        //             box_index = (i/3 + j/3) + (i/3 * 2);
        //             num_in_row = m["R"+to_string(i)].find(board[i][j]) != m["R"+to_string(i)].end();
        //             num_in_col = m["C"+to_string(j)].find(board[i][j]) != m["C"+to_string(j)].end();
        //             num_in_box = m["B"+to_string(box_index)].find(board[i][j]) != m["B"+to_string(box_index)].end();
        //             if (num_in_row || num_in_col || num_in_box) {
        //                 return false;
        //             }
        //             m["R"+to_string(i)].insert(board[i][j]);
        //             m["C"+to_string(j)].insert(board[i][j]);
        //             m["B"+to_string(box_index)].insert(board[i][j]);
        //         }
        //     }
        // }
        // return true;
    }
};

File: LeetCode/36ValidSudoku/ValidSudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        # BOTH SOLUTIONS HAVE SAME TIME COMPLEXITY, BUT ARRAYS IS MUCH FASTER

        # Arrays -> Time:O(n^2), Space:O(n^2) n=SideOfBoard
        rows = [[0 for i in range(9)] for i in range(9)]
        cols = [[0 for i in range(9)] for i in range(9)]
        boxes = [[0 for i in range(9)] for i in range(9)]
        for i in range(9):
            for j in range(9):
                if board[i][j] != '.':
                    box_index = (i//3 + j//3) + (i//3 * 2)
                    num = int(board[i][j]) - 1
                    if rows[i][num] \
                    or cols[num][j] \
                    or boxes[box_index][num]:
                        return False
                    rows[i][num] = 1
                    cols[num][j] = 1
                    boxes[box_index][num] = 1
        return True


        # # Dictionary Set -> Time:O(n^2), Space:O(n^2) n=SideOfBoard
        # d = {}
        # for i in range(len(board)):
        #     d[f'R{i}'] = set()
        #     d[f'C{i}'] = set()
        #     d[f'B{i}'] = set()
        # for i in range(9):
        #     for j in range(9):
        #         box_index = (i//3 + j//3) + (i//3 * 2)
        #         if board[i][j] != '.':
        #             if board[i][j] in d[f'R{i}'] \
        #             or board[i][j] in d[f'C{j}'] \
        #             or board[i][j] in d[f'B{box_index}']:
        #                 return False
        #             d[f'R{i}'].add(board[i][j])
        #             d[f'C{j}'].add(board[i][j])
        #             d[f'B{box_index}'].add(board[i][j])
        # return True
